// @opensouls/pipeline/dist/index.mjs
// src/filePipeline.ts
import { glob } from "glob";
import { join, relative } from "node:path";
var filePathToKey = (path) => {
  return path.replace(/[\\\/]/g, "__").replace(/[^\w\d_\.]/g, "-");
};
var normalizeProcessCallbackReturn = (relativePath, result) => {
  if (typeof result === "string") {
    return [{ content: result, key: filePathToKey(relativePath) }];
  }
  return result.map((item, index) => {
    if (typeof item === "string") {
      return { content: item, key: filePathToKey(relativePath) + "_" + index };
    }
    return { content: item.content, key: item.key ?? filePathToKey(relativePath) + "_" + index };
  });
};
var FilePipeline = class {
  constructor(src, dest, opts = {}) {
    this.src = src;
    this.dest = dest;
    this.opts = opts;
  }
  /**
   * Processes each file in the source directory, applying a provided callback function to transform the file content.
   * Each file is read and provided to the callback in two forms: as a UTF-8 string and as raw bytes.
   * The callback can return either a string, an array of strings, or an array of objects containing the content and an optional key.
   * If a key is not provided, a default key is generated based on the file's relative path and an index (if needed).
   * The transformed content is then written to the destination directory under the generated or provided key.
   * 
   * @param callback - A function that takes a `CallbackParams` object and returns a `ProcessCallbackReturn`.
   *                   This function is expected to perform the necessary transformations on the file content.
   * @returns - A promise that resolves when all files have been processed.
   */
  async process(callback) {
    const { mkdir, readFile, writeFile, stat } = await import("node:fs/promises");
    const { emptyDir } = await import("fs-extra/esm");
    let globSrc = join(this.src, "**/*");
    if (this.opts.replace) {
      await emptyDir(this.dest);
    } else {
      await mkdir(this.dest, { recursive: true });
    }
    const files = await glob(globSrc, { absolute: true });
    for (const filePath of files) {
      if ((await stat(filePath)).isDirectory()) {
        continue;
      }
      const relativeToSrc = relative(this.src, filePath);
      console.log(`processing ${relativeToSrc}`);
      const content = () => {
        return readFile(filePath, "utf8");
      };
      const contentBytes = () => {
        return readFile(filePath);
      };
      const result = await callback({
        content,
        contentBytes,
        path: relativeToSrc
      });
      const normalizedResult = normalizeProcessCallbackReturn(relativeToSrc, result);
      for (const { content: content2, key } of normalizedResult) {
        const destPath = join(this.dest, key);
        console.log(`writing ${destPath}`);
        await writeFile(destPath, content2);
      }
    }
  }
};
export {
  FilePipeline,
  filePathToKey
};
